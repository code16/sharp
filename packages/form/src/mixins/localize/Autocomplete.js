import localizeField from './field';
import { isLocaleObject } from "../../util/locale";

export default {
    mixins: [localizeField],
    computed: {
        localizedSearchKeys() {
            return this.localized
                ? this.searchKeys.map(key => {
                    let res = key;
                    if(this.localizedDataKeys.includes(key))
                        res+=`.${this.locale}`;
                    return res;
                })
                : this.searchKeys;
        },
        localizedDataKeys() {
            return (Array.isArray(this.localValues) && this.localValues.length
                ? Object.keys(this.localValues[0]).filter(key => this.isLocaleObject(this.localValues[0][key]))
                : []);
        }
    },
    methods: {
        isLocaleObject(obj) {
            return this.locales && isLocaleObject(obj, this.locales);
        },
        localizedTemplateData(templateData) {
            return this.localized ? Object.entries(templateData).reduce((res, [key, value]) => {
                res[key] = this.isLocaleObject(value) ? value[this.locale] : value;
                return res;
            }, {}) : templateData;
        }
    }
}