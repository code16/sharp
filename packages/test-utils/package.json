{
    "name": "@sharp/test-utils",
    "private": true,
    "main": "src/index.js",
    "version": "1.0.0"
}