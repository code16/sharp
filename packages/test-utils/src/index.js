export { default as MockInjections } from './MockInjections';
export { default as MockI18n } from './MockI18n';
export { default as QueryComponent } from './QueryComponent';
export { default as MockTransitions } from './MockTransitions';
export { mockSFC, unmockSFC } from './mockSFC';
export { mockProperty, unmockProperty, setter } from './mock-utils';
export { nextRequestFulfilled } from './moxios-utils';

export function wait(delay) {
    return new Promise(resolve => {
        setTimeout(resolve, delay);
    });
}