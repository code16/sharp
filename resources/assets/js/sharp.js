import './polyfill';
import Vue from 'vue';
import Vuex from 'vuex';
import VueRouter from 'vue-router';
import Trix from 'trix';
import { install as VueGoogleMaps } from 'vue2-google-maps';
import Notifications from 'vue-notification';
import locale from 'element-ui/lib/locale';
import { elLang } from './util/element-ui';


import SharpCommands from 'sharp-commands';
import SharpDashboard from 'sharp-dashboard';
import SharpEntityList from 'sharp-entity-list';
import SharpFilters from 'sharp-filters';
import SharpForm from 'sharp-form';
import SharpShow from 'sharp-show';
import SharpUI from 'sharp-ui';

import ActionView from './components/ActionView';
import LeftNav from './components/LeftNav';
import {
    ItemVisual,
    CollapsibleItem,
    NavItem,
} from 'sharp-ui';

import { store as getStore } from './store/store';
import { router as getRouter } from "./router";

locale.use(elLang());

Vue.use(Notifications);
Vue.use(VueGoogleMaps, {
    installComponents: false
});

Vue.use(VueRouter);
Vue.use(Vuex);

const router = getRouter();
const store = getStore();

Vue.use(SharpCommands, { store, router });
Vue.use(SharpDashboard, { store, router });
Vue.use(SharpEntityList, { store, router });
Vue.use(SharpFilters, { store, router });
Vue.use(SharpForm, { store, router });
Vue.use(SharpShow, { store, router });
Vue.use(SharpUI, { store, router });

window.Trix = Trix;
Vue.config.ignoredElements = [/^trix-/];


new Vue({
    el: "#sharp-app",

    components: {
        'sharp-action-view': ActionView,
        'sharp-left-nav': LeftNav,
        'sharp-collapsible-item': CollapsibleItem,
        'sharp-nav-item': NavItem,
        'sharp-item-visual': ItemVisual
    },

    created() {
        this.$on('setClass',(className,active)=> {
            this.$el.classList[active ? 'add' : 'remove'](className);
        });
        if(this.$route.query['x-access-from']) {
            this.$router.replace({
                ...this.$route,
                query: {
                    ...this.$route.query,
                    'x-access-from': undefined
                }
            });
        }
    },

    store,
    router,
});




