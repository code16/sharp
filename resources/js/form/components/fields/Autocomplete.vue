<script setup lang="ts">
    import FormFieldLayout from "@/form/components/FormFieldLayout.vue";
    import { FormFieldEmits, FormFieldProps } from "@/form/types";
    import {
        FormAutocompleteItemData,
        FormAutocompleteLocalFieldData,
        FormAutocompleteRemoteFieldData,
    } from "@/types";
    import { Popover, PopoverContent, PopoverTrigger } from "@/components/ui/popover";
    import { computed, ref } from "vue";
    import { Button } from "@/components/ui/button";
    import { __, trans_choice } from "@/utils/i18n";
    import { ChevronsUpDown, X, Check } from 'lucide-vue-next';
    import {
        Command,
        CommandEmpty,
        CommandGroup,
        CommandInput, CommandItem,
        CommandList,
    } from "@/components/ui/command";
    import { route } from "@/utils/url";
    import { api } from "@/api/api";
    import { useParentForm } from "@/form/useParentForm";
    import { isCancel } from "axios";
    import { ComboboxItemIndicator } from "reka-ui";
    import { useParentCommands } from "@/commands/useCommands";
    import { useIsInDialog } from "@/components/ui/dialog/Dialog.vue";
    import { useFullTextSearch } from "@/composables/useFullTextSearch";

    const props = defineProps<FormFieldProps<FormAutocompleteLocalFieldData | FormAutocompleteRemoteFieldData>>();
    const emit = defineEmits<FormFieldEmits<FormAutocompleteLocalFieldData | FormAutocompleteRemoteFieldData>>();
    const form = useParentForm();

    const open = ref(false);
    const searchTerm = ref('');
    const results = ref<FormAutocompleteItemData[]>([]);
    const loading = ref(false);

    let abortController: AbortController | null = null;
    let timeout = null;
    let loadingTimeout = null;

    const parentCommands = useParentCommands();
    const isInDialog = useIsInDialog();
    const { fullTextSearch } = useFullTextSearch(
        () => props.field.mode === 'local' ? props.field.localValues : null,
        {
            id: props.field.itemIdAttribute,
            searchKeys: props.field.mode === 'local' ? props.field.searchKeys : [],
        }
    );

    function search(query: string, immediate?: boolean) {
        if(props.field.mode === 'remote') {
            const field = props.field as FormAutocompleteRemoteFieldData;
            clearTimeout(timeout);
            if(query.length >= field.searchMinChars) {
                if(!results.value.length) {
                    loading.value = true;
                }
                if(immediate) {
                    remoteSearch(query);
                } else {
                    timeout = setTimeout(() => remoteSearch(query), props.field.debounceDelay)
                }
            } else {
                clearTimeout(timeout);
                loading.value = false;
                results.value = [];
            }
        } else {
            results.value = !query.length
                ? props.field.localValues
                : fullTextSearch(query);
        }
    }

    async function remoteSearch(query: string) {
        const field = props.field as FormAutocompleteRemoteFieldData;
        clearTimeout(loadingTimeout);
        loadingTimeout = setTimeout(() => {
            loading.value = true;
        }, 200);
        abortController?.abort();
        abortController = new AbortController();
        results.value = await api.post(
            route('code16.sharp.api.form.autocomplete.index', {
                entityKey: form.entityKey,
                autocompleteFieldKey: props.parentField ? `${props.parentField.key}.${field.key}` : field.key,
                embed_key: form.embedKey,
                entity_list_command_key: parentCommands?.commandContainer === 'entityList' ? form.commandKey : null,
                show_command_key: parentCommands?.commandContainer === 'show' ? form.commandKey : null,
                instance_id: form.instanceId,
                endpoint: field.remoteEndpoint,
                search: query,
            }), {
                formData: field.callbackLinkedFields
                    ? Object.fromEntries(
                        Object.entries(props.parentData).filter(([fieldKey]) => field.callbackLinkedFields.includes(fieldKey))
                    )
                    : null,
            }, {
                signal: abortController.signal,
            }
        )
            .then(response => {
                clearTimeout(loadingTimeout);
                loading.value = false;
                return response;
            }, (e) => {
                if(isCancel(e)) {
                    clearTimeout(loadingTimeout);
                }
                return Promise.reject(e);
            })
            .then(response => response.data.data)
        ;
    }

    let hasTyped = false;

    function onSearchInput(query: string) {
        if(!query.length && !hasTyped) {
            return;
        }
        hasTyped = true;
        search(query);
    }

    function onOpen() {
        if(!searchTerm.value && (props.field.mode === 'local' || props.field.searchMinChars === 0)) {
            search('', true);
        }
    }

    function onSelect(id: string | number) {
        emit('input', results.value.find(r => String(r[props.field.itemIdAttribute]) === String(id)));
        open.value = false;
    }

    const currentLocalValue = computed(() => {
        return props.field.mode === 'local' && props.value
            ? props.field.localValues?.find(v => String(props.value[props.field.itemIdAttribute]) === String(v[props.field.itemIdAttribute]))
            : null;
    })
</script>

<template>
    <FormFieldLayout v-bind="props" field-group v-slot="{ ariaLabelledBy, ariaDescribedBy }">
        <Popover v-model:open="open" @update:open="$event ? onOpen() : null">
            <template v-if="props.value">
                <div class="relative">
                    <PopoverTrigger as-child>
                        <div class="relative border border-input flex items-center rounded-md min-h-10 text-sm px-3 py-2 pr-10 bg-background focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 aria-disabled:pointer-events-none aria-disabled:opacity-50"
                            role="combobox"
                            aria-autocomplete="none"
                            tabindex="0"
                            :aria-labelledby="ariaLabelledBy"
                            :aria-describedby="ariaDescribedBy"
                            :aria-disabled="props.field.readOnly"
                        >
                            <div class="flex-1"
                                v-html="currentLocalValue
                                    ? currentLocalValue._htmlResult ?? currentLocalValue._html
                                    : props.value._htmlResult ?? props.value._html ?? props.value[props.field.itemIdAttribute]
                                "
                            ></div>
                        </div>
                    </PopoverTrigger>
                    <Button class="absolute right-0 h-9.5 top-1/2 -translate-y-1/2 opacity-50 hover:opacity-100"
                        :disabled="props.field.readOnly"
                        variant="ghost"
                        size="icon"
                        :aria-label="__('sharp::form.select.clear')"
                        @click="$emit('input', null)"
                    >
                        <X class="size-4" aria-hidden="true" />
                    </Button>
                </div>
            </template>
            <template v-else>
                <PopoverTrigger as-child>
                    <Button
                        class="w-full justify-between text-muted-foreground px-3"
                        variant="outline"
                        role="combobox"
                        aria-autocomplete="none"
                        :disabled="props.field.readOnly"
                        :aria-labelledby="ariaLabelledBy"
                        :aria-describedby="ariaDescribedBy"
                    >
                        {{ props.field.placeholder ?? __('sharp::form.autocomplete.placeholder') }}
                        <ChevronsUpDown class="ml-2 h-4 w-4 shrink-0 opacity-50 text-foreground" />
                    </Button>
                </PopoverTrigger>
            </template>

            <PopoverContent
                class="p-0 w-(--reka-popover-trigger-width) min-w-[200px]"
                align="start"
                :avoid-collisions="false"
            >
                <Command
                    :class="isInDialog ? 'max-h-(--reka-popper-available-height)' : ''"
                    :model-value="value?.[props.field.itemIdAttribute]"
                    :reset-search-term-on-blur="false"
                    ignore-filter
                    @update:model-value="onSelect($event as any)"
                >
                    <CommandInput
                        v-model="searchTerm"
                        @update:model-value="onSearchInput"
                        :display-value="() => searchTerm"
                        :placeholder="props.value ? props.field.placeholder ?? __('sharp::form.autocomplete.placeholder') : null"
                    />
                    <CommandList>
                        <template v-if="loading">
                            <div class="py-6 px-4 text-center text-sm">
                                {{ __('sharp::form.autocomplete.loading') }}
                            </div>
                        </template>
                        <template v-else-if="!results?.length && props.field.mode === 'remote' && searchTerm.length < props.field.searchMinChars">
                            <div class="py-6 px-4 text-center text-sm">
                                {{ trans_choice('sharp::form.autocomplete.query_too_short', props.field.searchMinChars, { min_chars: props.field.searchMinChars }) }}
                            </div>
                        </template>
                        <template v-else>
                            <CommandEmpty>
                                {{ __('sharp::form.autocomplete.no_results_text') }}
                            </CommandEmpty>
                            <CommandGroup v-show="results.length">
                                <template v-for="item in results" :key="item[props.field.itemIdAttribute]">
                                    <CommandItem
                                        class="group/item"
                                        :value="item[props.field.itemIdAttribute]"
                                    >
                                        <div class="size-4">
                                            <ComboboxItemIndicator as-child>
                                                <Check class="size-4" />
                                            </ComboboxItemIndicator>
                                        </div>

                                        <div v-html="item._html"></div>
                                    </CommandItem>
                                </template>
                            </CommandGroup>
                        </template>
                    </CommandList>
                </Command>
            </PopoverContent>
        </Popover>
    </FormFieldLayout>
</template>
