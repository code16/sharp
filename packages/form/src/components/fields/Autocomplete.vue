<template>
    <div class="SharpAutocomplete" :class="classes">
        <template v-if="ready">
            <Multiselect
                class="SharpAutocomplete__multiselect"
                :class="{ 'SharpAutocomplete__multiselect--hide-dropdown':hideDropdown }"
                :value="value"
                :options="suggestions"
                :track-by="itemIdAttribute"
                :internal-search="false"
                :placeholder="placeholder"
                :loading="isLoading"
                :multiple="multiple"
                :disabled="readOnly"
                :hide-selected="hideSelected"
                :allow-empty="allowEmpty"
                :preserve-search="preserveSearch"
                :show-pointer="showPointer"
                :searchable="searchable"
                @search-change="updateSuggestions($event)"
                @select="handleSelect"
                @input="$emit('multiselect-input',$event)"
                @close="handleDropdownClose"
                @open="handleDropdownOpen"
                ref="multiselect"
            >
                <template v-slot:clear>
                    <template v-if="clearButtonVisible">
                        <button class="SharpAutocomplete__result-item__close-button" type="button" @click="handleClearButtonClicked">
                            <svg class="SharpAutocomplete__result-item__close-icon"
                                aria-label="close" width="10" height="10" viewBox="0 0 10 10" fill-rule="evenodd">
                                <path d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"></path>
                            </svg>
                        </button>
                    </template>
                </template>
                <template v-slot:singleLabel="{ option }">
                    <TemplateRenderer
                        name="ResultItem"
                        :template="resultItemTemplate"
                        :template-data="localizedTemplateData(option)"
                    />
                </template>
                <template v-slot:option="{ option }">
                    <TemplateRenderer
                        name="ListItem"
                        :template="listItemTemplate"
                        :template-data="localizedTemplateData(option)"
                    />
                </template>
                <template v-slot:loading>
                    <Loading :visible="isLoading" small />
                </template>
                <template v-slot:noResult>
                    {{ l('form.autocomplete.no_results_text') }}
                </template>
            </multiselect>

            <template v-if="overlayVisible">
                <div class="SharpAutocomplete__overlay multiselect">
                    <div class="multiselect__tags">
                        <TemplateRenderer
                            name="ResultItem"
                            :template="resultItemTemplate"
                            :template-data="localizedTemplateData(value)"
                        />
                    </div>
                </div>
            </template>
        </template>
    </div>
</template>

<script>
    import debounce from 'lodash/debounce';
    import Multiselect from 'vue-multiselect';
    import { warn, lang, search } from 'sharp';
    import { TemplateRenderer } from 'sharp/components';
    import { Loading } from 'sharp-ui';
    import { Localization } from 'sharp/mixins';

    import { getAutocompleteSuggestions } from "../../api";
    import localize from '../../mixins/localize/Autocomplete';
    import { setDefaultValue } from "../../util";


    export default {
        name:'SharpAutocomplete',
        components: {
            Multiselect,
            TemplateRenderer,
            Loading,
        },

        mixins: [Localization, localize],

        props: {
            fieldKey: String,

            value: [String, Number, Object, Array],

            mode: String,
            localValues: {
                type: Array,
                default:()=>[]
            },
            placeholder: {
                type: String,
                default: () => lang('form.multiselect.placeholder')
            },
            remoteEndpoint: String,
            remoteMethod: String,
            remoteSearchAttribute: {
                type: String,
                default: 'query'
            },
            itemIdAttribute: {
                type:String,
                default: 'id'
            },
            searchMinChars: {
                type: Number,
                default: 1
            },
            searchKeys: {
                type: Array,
                default:()=>['value']
            },
            dataWrapper: String,
            readOnly: Boolean,
            listItemTemplate: String,
            resultItemTemplate: String,
            noResultItem: Boolean,
            multiple: Boolean,
            hideSelected: Boolean,
            searchable: {
                type: Boolean,
                default: true,
            },
            allowEmpty: {
                type: Boolean,
                default: true
            },
            clearOnSelect: Boolean,
            preserveSearch: {
                type: Boolean,
                default: true
            },
            showPointer: {
                type:Boolean,
                default:true
            },
            dynamicAttributes: Array,
        },
        data() {
            return {
                ready: false,
                query: '',
                suggestions: this.localValues,
                opened: false,
                isLoading: false,
            }
        },
        watch: {
            localValues() {
                if(!this.isRemote) {
                    this.updateLocalSuggestions(this.query);
                }
            },
        },
        computed: {
            isRemote() {
                return this.mode === 'remote';
            },
            hideDropdown() {
                return this.isQueryTooShort;
            },
            isQueryTooShort() {
                return this.isRemote && this.query.length < this.searchMinChars;
            },
            clearButtonVisible() {
                return !!this.value && !this.opened;
            },
            classes() {
                return [
                    { 'SharpAutocomplete--remote': this.isRemote },
                    { 'SharpAutocomplete--disabled': this.readOnly }
                ];
            },
            overlayVisible() {
                const isFormField = !!this.fieldKey;
                return this.value && isFormField;
            }
        },
        methods: {
            updateSuggestions(query) {
                this.query = query;
                if(this.isQueryTooShort) {
                    return;
                }
                if(this.isRemote) {
                    this.isLoading = true;
                    this.updateRemoteSuggestions(query);
                }
                else {
                    this.updateLocalSuggestions(query);
                }
            },

            updateLocalSuggestions(query) {
                this.suggestions = query.length >= this.searchMinChars
                    ? search(this.localValues, query, { searchKeys: this.searchKeys })
                    : this.localValues;
            },
            updateRemoteSuggestions: debounce(function(query) {
                return getAutocompleteSuggestions({
                    url: this.remoteEndpoint,
                    method: this.remoteMethod,
                    locale: this.locale,
                    searchAttribute: this.remoteSearchAttribute,
                    dataWrapper: this.dataWrapper,
                    fieldKey: this.fieldKey,
                    query,
                })
                .then(suggestions => {
                    this.suggestions = suggestions;
                })
                .finally(() => {
                    this.isLoading = false;
                });
            }, 200),

            handleSelect(value) {
                this.$emit('input', value);
            },
            handleDropdownClose() {
                this.opened = false;
                this.$emit('close');
            },
            handleDropdownOpen() {
                this.opened = true;
                this.$emit('open');
            },
            handleClearButtonClicked() {
                this.$emit('input', null);
                this.$nextTick(() => {
                    this.$refs.multiselect.activate();
                });
            },
            itemMatchValue(localValue) {
                // noinspection EqualityComparisonWithCoercionJS
                return localValue[this.itemIdAttribute] == this.value[this.itemIdAttribute];
            },
            findLocalValue() {
                if(!this.value || this.value[this.itemIdAttribute] == null) return null;
                if(!this.localValues.some(this.itemMatchValue)) {
                    error(`Autocomplete (key: ${this.fieldKey}) can't find local value matching : ${JSON.stringify(this.value)}`);
                    return null;
                }
                return this.localValues.find(this.itemMatchValue);
            },
            async setDefault() {
                this.$emit('input', this.findLocalValue(), { force: true });
                await this.$nextTick();
                this.ready = true;
            }
        },
        created() {
            if(this.mode === 'local' && !this.searchKeys) {
                warn(`Autocomplete (key: ${this.fieldKey}) has local mode but no searchKeys, default set to ['value']`);
            }
            if(this.isRemote) {
                this.ready = true;
            } else {
                setDefaultValue(this, this.setDefault, {
                    dependantAttributes: ['localValues'],
                });
            }
        }
    }
</script>