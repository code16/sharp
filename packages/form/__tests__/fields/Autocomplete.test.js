import Vue from 'vue';
import Autocomplete from '../../src/components/fields/Autocomplete.vue';

import { MockI18n, nextRequestFulfilled } from '@sharp/test-utils';
import moxios from 'moxios';
import { search } from 'sharp';

jest.mock('sharp/util/search', ()=>({
    search: jest.fn(()=>[]),
}));

describe('autocomplete-field', ()=>{
    Vue.use(MockI18n);

    describe('common & local', ()=> {
        beforeEach(()=>{
            document.body.innerHTML = `
            <div id="app">
                <sharp-autocomplete :value="value" 
                                    :read-only="readOnly"
                                    mode="local"
                                    placeholder="nom"
                                    item-id-attribute="id" 
                                    :search-keys="['name', 'alias']"
                                    :list-item-template="'<em>{{name}}, {{alias}}</em>'"
                                    :result-item-template="'<b>{{name}}, {{alias}}</b>'"
                                    :local-values="[{id:1, name:'Theodore Bagwell', alias:'T-Bag'},
                                                    {id:2, name:'Lincoln Burrows', alias: 'Linc'}]"
                                    @input="inputEmitted($event)"
                                    >
                </sharp-autocomplete>
            </div>`;
            MockI18n.mockLangFunction();
        });

        test('can mount Autocomplete field as multiselect', async () => {
            await createVm();

            expect(document.body.innerHTML).toMatchSnapshot();
        });

        test('can mount Autocomplete field as result item', async () => {
            await createVm({
                data: () => ({ value: {id:1, name:'Theodore Bagwell', alias:'T-Bag'} })
            });

            expect(document.body.innerHTML).toMatchSnapshot();
        });

        test('expose appropriate props to multiselect component', async ()=>{
            let $autocomplete = await createVm({
                propsData: {
                    readOnly: true
                },
                data: () => ({ value: null })
            });

            let { multiselect } = $autocomplete.$refs;

            expect(multiselect.$props).toMatchObject({
                options: [{id: 1, name:'Theodore Bagwell', alias: 'T-Bag'},
                    {id: 2, name:'Lincoln Burrows', alias: 'Linc'}],
                value: null,
                multiple: false,
                trackBy: 'id',
                searchable: true,
                clearOnSelect: true,
                hideSelected: false,
                placeholder: 'nom',
                allowEmpty: true,
                resetAfter: false,
                closeOnSelect: true,
                taggable: false,
                preserveSearch: true,
                internalSearch: false,

                disabled: true,
            });
        });

        test('search', async () => {
            let $autocomplete = await createVm();

            let { multiselect } = $autocomplete.$refs;

            multiselect.$emit('search-change', 'query');

            expect(search).toHaveBeenCalledWith(
                [{id:1, name:'Theodore Bagwell', alias:'T-Bag'}, {id:2, name:'Lincoln Burrows', alias: 'Linc'}],
                'query',
                {
                    searchKeys: ['name', 'alias']
                }
            );

            // expect($autocomplete.searchStrategy).toMatchObject({
            //     fuse: {
            //         list: [{id:1, name:'Theodore Bagwell', alias:'T-Bag'}, {id:2, name:'Lincoln Burrows', alias: 'Linc'}]
            //     },
            //     options: {
            //         keys: ['name', 'alias']
            //     }
            // });
        });


        test('sync opened with multiselect', async () => {
            let $autocomplete = await createVm({
                data: () => ({ value:null })
            });
            let { multiselect } = $autocomplete.$refs;

            expect($autocomplete.opened).toBeFalsy();

            multiselect.$emit('open');

            expect($autocomplete.opened).toBe(true);

            multiselect.$emit('close');

            expect($autocomplete.opened).toBe(false);
        });

        test('emit input on select & correct value', async () => {
            let inputEmitted = jest.fn();

            let $autocomplete = await createVm({
                data: () => ({ value:null }),
                methods: {
                    inputEmitted
                }
            });
            let { $root:vm } = $autocomplete;
            let { multiselect } = $autocomplete.$refs;

            multiselect.$emit('select', {id:2, name:'Lincoln Burrows', alias: 'Linc'});

            expect(inputEmitted).toHaveBeenLastCalledWith({id:2, name:'Lincoln Burrows', alias: 'Linc'});
        });

        test('define loading slot', async () => {
            let $autocomplete = await createVm();
            let { multiselect } = $autocomplete.$refs;
            expect(multiselect.$slots.loading).toBeTruthy();
        });

        test('define noResult slot', async () => {
            let $autocomplete = await createVm();
            let { multiselect } = $autocomplete.$refs;
            expect(multiselect.$slots.loading).toBeTruthy();
        });
    });

    describe('remote', () => {
        beforeEach(()=>{
            document.body.innerHTML = `
            <div id="app">
                <sharp-autocomplete 
                    :value="value" 
                    :read-only="readOnly" 
                    mode="remote"
                    placeholder="nom"
                    item-id-attribute="id" 
                    remote-endpoint="/autocomplete"
                    :search-min-chars="3"
                    :remote-method="remoteMethod || 'GET'"
                    remote-search-attribute="search"
                    :list-item-template="'<em>{{name}}, {{alias}}</em>'"
                    :result-item-template="'<b>{{name}}, {{alias}}</b>'"
                    @input="inputEmitted($event)"
                />
            </div>`;
            moxios.install();
        });

        afterEach(()=>{
            moxios.uninstall();
        });

        test("'loading' computed property changes properly", async () => {
            let $autocomplete = await createVm({
                propsData: {
                    remoteMethod: 'POST'
                }
            });

            let { multiselect } = $autocomplete.$refs;

            expect($autocomplete.isLoading).toBeFalsy();

            multiselect.$emit('open');
            multiselect.$emit('search-change','Li'); // less than min chars
            expect($autocomplete.isLoading).toBe(false);

            multiselect.$emit('search-change','Linc');
            expect($autocomplete.isLoading).toBe(true);

            await nextRequestFulfilled({ status: 200, response:[] }, 210);

            expect($autocomplete.isLoading).toBe(false);
        });

        test("'hide dropdown' computed property changes properly (query less than min chars)", async () => {
            let $autocomplete = await createVm();

            let { multiselect } = $autocomplete.$refs;

            expect($autocomplete.hideDropdown).toBe(true);

            multiselect.$emit('search-change', 'Th');
            expect($autocomplete.hideDropdown).toBe(true);

            multiselect.$emit('search-change', 'The');
            expect($autocomplete.hideDropdown).toBe(false);

        });

        test('Load response data', async () => {
            let $autocomplete = await createVm();

            let { multiselect } = $autocomplete.$refs;

            multiselect.$emit('search-change', 'Theo');
            await nextRequestFulfilled({ status:200, response:[{id:1, name:'Theodore Bagwell', alias:'T-Bag'}] }, 210);

            expect($autocomplete.suggestions).toEqual([{id:1, name:'Theodore Bagwell', alias:'T-Bag'}]);
        });

        test("Send proper GET request", async () => {
            let $autocomplete = await createVm({
                propsData: {
                    remoteMethod: 'GET'
                }
            });

            let { multiselect } = $autocomplete.$refs;

            multiselect.$emit('search-change', 'Linc');

            let { request } = await nextRequestFulfilled({ status:200, response:[] }, 210);
            let { config:{ method, params, url }} = request;

            expect(method).toBe('get');
            expect(params).toEqual({ search: 'Linc' });
            expect(url).toBe('/autocomplete');
        });

        test("Send proper POST request", async () => {
            let $autocomplete = await createVm({
                propsData: {
                    remoteMethod: 'POST'
                }
            });

            let { multiselect } = $autocomplete.$refs;

            multiselect.$emit('search-change', 'Linc');

            let { request } = await nextRequestFulfilled({ status:200, response:[] }, 400);
            let { config:{ method, data, url }} = request;

            expect(method).toBe('post');
            expect(JSON.parse(data)).toEqual({ search: 'Linc' });
            expect(url).toBe('/autocomplete');
        });
    });

});

async function createVm(customOptions={}) {
    const vm = new Vue({
        el: '#app',
        mixins: [customOptions],

        components: {
            'sharp-autocomplete': Autocomplete
        },

        props:['readOnly', 'remoteMethod'],

        'extends': {
            data:()=>({ value: null }),
            methods: {
                inputEmitted:()=>{}
            }
        }
    });

    await Vue.nextTick();

    return vm.$children[0];
}