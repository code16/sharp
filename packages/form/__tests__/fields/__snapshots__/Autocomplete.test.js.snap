// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autocomplete-field common & local can mount Autocomplete field as multiselect 1`] = `
<div id="app">
	<div class="SharpAutocomplete">
		<div tabindex="-1" class="multiselect SharpAutocomplete__multiselect">
			<div class="multiselect__select"></div>
			<div class="multiselect__tags">
				<div class="multiselect__tags-wrap" style="display: none;"></div>
				<!---->
				<div class="SharpLoading SharpLoading--small" style="display: none;">
					<div class="SharpLoading__content">
						<svg viewBox="-75 -75 150 150" class="SharpLoading__svg">
							<circle cx="0" cy="0" r="37.5"></circle>
						</svg>
					</div>
				</div>
				<input name="" type="text" autocomplete="off" placeholder="nom" tabindex="0" class="multiselect__input" style="width: 0px; position: absolute; padding: 0px;">
				<!---->
				<span class="multiselect__placeholder">
					nom
				</span>
			</div>
			<div tabindex="-1" class="multiselect__content-wrapper" style="max-height: 300px; display: none;">
				<ul class="multiselect__content" style="display: inline-block;">
					<!---->
					<li class="multiselect__element">
						<span data-select="Press enter to select" data-selected="Selected" data-deselect="Press enter to remove" class="multiselect__option multiselect__option--highlight">
							<div class="SharpTemplate">
								<em>Theodore Bagwell, T-Bag</em>
							</div>
						</span>
						<!---->
					</li>
					<li class="multiselect__element">
						<span data-select="Press enter to select" data-selected="Selected" data-deselect="Press enter to remove" class="multiselect__option">
							<div class="SharpTemplate">
								<em>Lincoln Burrows, Linc</em>
							</div>
						</span>
						<!---->
					</li>
					<li style="display: none;">
						<span class="multiselect__option">
							{{ form.autocomplete.no_results_text }}
						</span>
					</li>
					<li style="display: none;">
						<span class="multiselect__option">List is empty.</span>
					</li>
				</ul>
			</div>
		</div>
		<!---->
	</div>
</div>
`;

exports[`autocomplete-field common & local can mount Autocomplete field as result item 1`] = `
<div id="app">
	<div class="SharpAutocomplete">
		<div tabindex="-1" class="multiselect SharpAutocomplete__multiselect">
			<div class="multiselect__select"></div>
			<button type="button" class="SharpAutocomplete__result-item__close-button">
				<svg aria-label="close" width="10" height="10" viewBox="0 0 10 10" fill-rule="evenodd" class="SharpAutocomplete__result-item__close-icon">
					<path d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"></path>
				</svg>
			</button>
			<div class="multiselect__tags">
				<div class="multiselect__tags-wrap" style="display: none;"></div>
				<!---->
				<div class="SharpLoading SharpLoading--small" style="display: none;">
					<div class="SharpLoading__content">
						<svg viewBox="-75 -75 150 150" class="SharpLoading__svg">
							<circle cx="0" cy="0" r="37.5"></circle>
						</svg>
					</div>
				</div>
				<input name="" type="text" autocomplete="off" placeholder="nom" tabindex="0" class="multiselect__input" style="width: 0px; position: absolute; padding: 0px;">
				<span class="multiselect__single">
					<div class="SharpTemplate">
						<b>Theodore Bagwell, T-Bag</b>
					</div>
				</span>
				<!---->
			</div>
			<div tabindex="-1" class="multiselect__content-wrapper" style="max-height: 300px; display: none;">
				<ul class="multiselect__content" style="display: inline-block;">
					<!---->
					<li class="multiselect__element">
						<span data-select="Press enter to select" data-selected="Selected" data-deselect="Press enter to remove" class="multiselect__option multiselect__option--highlight multiselect__option--selected">
							<div class="SharpTemplate">
								<em>Theodore Bagwell, T-Bag</em>
							</div>
						</span>
						<!---->
					</li>
					<li class="multiselect__element">
						<span data-select="Press enter to select" data-selected="Selected" data-deselect="Press enter to remove" class="multiselect__option">
							<div class="SharpTemplate">
								<em>Lincoln Burrows, Linc</em>
							</div>
						</span>
						<!---->
					</li>
					<li style="display: none;">
						<span class="multiselect__option">
							{{ form.autocomplete.no_results_text }}
						</span>
					</li>
					<li style="display: none;">
						<span class="multiselect__option">List is empty.</span>
					</li>
				</ul>
			</div>
		</div>
		<!---->
	</div>
</div>
`;
